zeus/carbon/carbon.py

```diff
@@ -0,0 +1,95 @@
+"""Carbon intensity providers used for carbon-aware optimizers."""
+
+from __future__ import annotations
+
+import abc
+import json
+import logging
+import requests
+
+logger = logging.getLogger(__name__)
+
+
+def get_ip_lat_long() -> tuple[float, float]:
+    """Retrieve the latitude and longitude of the current IP position."""
+    try:
+        ip_url = "http://ipinfo.io/json"
+        resp = requests.get(ip_url)
+        loc = resp.json()["loc"]
+        lat, long = map(float, loc.split(","))
+        logger.info("Retrieved latitude and longitude: %s, %s", lat, long)
+        return lat, long
+    except requests.exceptions.RequestException as e:
+        logger.exception(
+            "Failed to retrieve current latitude and longitude of IP: %s", e
+        )
+        raise
+
+
+class CarbonIntensityNotFoundError(Exception):
+    """Exception when carbon intensity measurement could not be retrieved."""
+
+    def __init__(self, message: str) -> None:
+        """Initialize carbon not found exception."""
+        super().__init__(message)
+
+
+class CarbonIntensityProvider(abc.ABC):
+    """Abstract class for implementing ways to fetch carbon intensity."""
+
+    def __init__(
+        self,
+        location: tuple[float, float],
+        estimate: bool = False,
+        emission_factor_type: str = "direct",
+    ) -> None:
+        """Initializes carbon intensity provider location to the latitude and longitude of the input `location`.
+
+        Args:
+            location: tuple of latitude and longitude (latitude, longitude)
+            estimate: bool to toggle whether carbon intensity is estimated or not
+            emission_factor_type: emission factor to be measured (`direct` or `lifestyle`)
+        """
+        self.lat, self.long = location
+        self.estimate = estimate
+        self.emission_factor_type = emission_factor_type
+
+    @abc.abstractmethod
+    def get_current_carbon_intensity(self) -> float:
+        """Abstract method for fetching the current carbon intensity of the set location of the class."""
+        pass
+
+
+class ElectrictyMapsClient(CarbonIntensityProvider):
+    """Carbon Intensity Provider with ElectricityMaps API.
+
+    ElectricityMaps: https://www.electricitymaps.com/
+    ElectricityMaps API: https://static.electricitymaps.com/api/docs/index.html
+    ElectricityMaps GitHub: https://github.com/electricitymaps/electricitymaps-contrib
```

Comment:
I have a hunch that this may not be parsed as intended in mkdocs, but I can be wrong. Could you check the doc build result with `scripts/preview_docs.sh`?

Reply:
Yup you were right. Fixed it so now it looks prettier.